File: src/main/java/com/example/hoteljavafinal/gestionChambrev1/hotel/HotelManager.java
package com.example.hoteljavafinal.gestionChambrev1.hotel;

import com.example.hoteljavafinal.administration.users.RoleAPP;
import com.example.hoteljavafinal.gestionChambrev1.chambre.ChambreEntity;
import com.example.hoteljavafinal.gestionChambrev1.hotel.exceptions.DroitInsuffisantException;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;

import java.util.Date;
import java.util.List;

@NoArgsConstructor
@AllArgsConstructor
public class HotelManager {

    /**
     * **************************************
     * *                                    *
     * *            GET Method              *
     * *                                    *
     * **************************************
     */
    public List<ChambreEntity> chambresHotels;
    public List<ChambreEntity> getChambresDisponibles(Date dateDebut, RoleAPP roleAPP) {
        return null;
    }
    public List<ChambreEntity> getAllChambresDisponibleToday(RoleAPP roleAPP) {
        return null;
    }
    public List<ChambreEntity> getAllChambresDisponible(RoleAPP roleAPP, Date dateDebut, Date dateFin) {
        return null;
    }

    public List<ChambreEntity> reserverUneChambre(RoleAPP roleAPP, Date dateDebut, Date dateFin, ChambreEntity chambreEntity) throws DroitInsuffisantException {
        if (roleAPP != RoleAPP.MANAGER) {
            throw new DroitInsuffisantException();
        }
        return null;
    }

}

File: hotel_TP_CPP/gestionChambre/hotel/HotelManager.cpp
#include <vector>
#include <chrono>
#include <stdexcept>
#include <iostream>
#include "../Chambre/ChambreEntity.h"

class HotelManager {
public:
    HotelManager() = default;
    HotelManager(const std::vector<ChambreEntity>& chambres) : chambresHotels(chambres) {}

    std::vector<ChambreEntity> chambresHotels;

    std::vector<ChambreEntity> getChambresDisponibles(std::chrono::system_clock::time_point dateDebut, RoleAPP roleAPP) {
        // Implémentation ici
        return {};
    }

    std::vector<ChambreEntity> getAllChambresDisponibleToday(RoleAPP roleAPP) {
        // Implémentation ici
        return {};
    }

    std::vector<ChambreEntity> getAllChambresDisponible(RoleAPP roleAPP, std::chrono::system_clock::time_point dateDebut, std::chrono::system_clock::time_point dateFin) {
        // Implémentation ici
        return {};
    }

    std::vector<ChambreEntity> reserverUneChambre(RoleAPP roleAPP, std::chrono::system_clock::time_point dateDebut, std::chrono::system_clock::time_point dateFin, ChambreEntity chambreEntity) {
        // Vérifier les droits de l'utilisateur
        if (!roleAPP.hasReservationRight()) {
            throw std::runtime_error("Droits insuffisants pour effectuer cette opération");
        }
        // Ajouter la réservation ici
        return {};
    }
};

int main() {
    // Utilisation de la classe HotelManager
    HotelManager manager;
    try {
        auto chambres = manager.reserverUneChambre(RoleAPP(), std::chrono::system_clock::now(), std::chrono::system_clock::now() + std::chrono::hours(24), ChambreEntity());
        std::cout << "Réservation réussie!" << std::endl;
    } catch (const std::runtime_error& e) {
        std::cout << "Erreur: " << e.what() << std::endl;
    }

    return 0;
}
